crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -425,13 +514,38 @@ unsafe impl<T: Resource> SystemParamState for ResMutState<T> {
         let archetype_component_id = resource_archetype
             .get_archetype_component_id(component_id)
             .unwrap();
-        system_meta
-            .archetype_component_access
-            .add_write(archetype_component_id);
-        Self {
+
+        let state = Self {
+            archetype_component_id,
             component_id,
             marker: PhantomData,
-        }
+        };
+        assert_component_access_compatibility(
+            system_meta,
+            std::any::type_name::<ResMut<T>>(),
+            &state,
+            world,
+            false,
+        );
+        system_meta
+            .component_access_set
+            .extend(state.component_access_set());
+        system_meta
+            .archetype_component_access
+            .extend(&state.archetype_component_access());
+        state
+    }
+
+    fn archetype_component_access(&self) -> Access<ArchetypeComponentId> {
```

Comment:
I thought about this for quite some while, I could not find another way to get the required information. In order for the `ParamSet` to accurately find and which parameters conflict with each other, atleast the `component_access_set` method has to be added. At that point it seems strange to not also have an `archetype_component_access`. But im open to suggestions.

Reply:
ah I didnt see your comment, reading through the branch now

Reply:
```rust
#(
    // Pretend to add each param to the system alone, see if it conflicts
    let mut #meta = system_meta.clone();
    #meta.component_access_set.clear();
    #meta.archetype_component_access.clear();
    #param_fetch::init(world, &mut #meta);
    let #param = #param_fetch::init(world, &mut system_meta.clone());
)*
```
Couldn't this have unexpected consequences? Is the operation of initializing a parameter idempotent? The performance shouldn't be an issue as you mentioned but it still seems like a bit of a hack. Suppose a parameter added an entry to a resource when it is initialized, this would break this implementation. But if this is clearly documented it might be worth it to simplify the trait as you say.

Reply:
I think it's fair to assume parameter initialization is idempotent, since the same parameter will be initialized every time it is used in a different system. However, I still feel like calling init multiple times for one usage is a bit of a hack.

Reply:
Yeah I do agree that double-inits are hack-ish, but so is double-defining accesses (which I think is a bigger threat to correctness at this stage).

I think I still prefer the double init approach (in the short term). I'm open to breaking out param access definitions into more structured methods, but only if they are treated as the single source of truth (and don't rely on specific init() impls handling that correctly). This would require rethinking SystemParam lifecycles so that we can call `SomeParam::get_archetype_component_access()` to register accesses in a way that is automatic and abstracted out from init impls.

For now, is anyone opposed to adopting double-init, then we can take our time to reconsider the SystemParam access api? Or alternatively, is someone willing to pick up that design work?